libraries/AP_HAL_PX4/UARTDriver.cpp

```diff
@@ -274,7 +275,10 @@ size_t PX4UARTDriver::write(uint8_t c)
         }
         hal.scheduler->delay(1);
     }
-    return _writebuf.write(&c, 1);
+	_write_sem.take(10);
```

Comment:
@OXINARF Oh sorry, just saw your top level comment actually. This was determined to be the problem because without semaphore writing data to a TeraRanger Tower at 3 chars per write didn't work, e.g. `uart->write("BBB");`. See here for a very rough version of what I was doing: https://github.com/ArduPilot/ardupilot/compare/master...danorbtnk:trtower_setupfix . 

I also had the guys from terabee look at it and they could find no issue with my code for configuring the tower.